python/ray/ml/utils/tensorflow_utils.py

```diff
@@ -0,0 +1,73 @@
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+
+
+def convert_pandas_to_tf_tensor(df: pd.DataFrame) -> tf.Tensor:
+    """Convert a pandas dataframe to a TensorFlow tensor.
+
+    This function works in two steps:
+    1. Convert each dataframe column to a tensor.
+    2. Concatenate the resulting tensors along the last axis.
+
+    Arguments:
+        df: The dataframe to convert to a TensorFlow tensor. Columns must be of
+            a numeric dtype, ``TensorDtype``, or object dtype. If a column has
+            an object dtype, the column must contain ``ndarray`` objects.
+
+    Returns:
+        A tensor constructed from the dataframe.
+
+    Raises:
+        ValueError: if a column has an invalid dtype.
+        ValueError: if the columns can't be combined into a single tensor.
+
+    Examples:
+        >>> import pandas as pd
+        >>> from ray.ml.utils.tensorflow_utils improt convert_pandas_to_tf_tensor
+        >>>
+        >>> df = pd.DataFrame({"X1": [1, 2, 3], "X2": [4, 5, 6]})
+        >>> convert_pandas_to_tf_tensor(df[["X1"]]).shape
+        TensorShape([3, 1])
+        >>> convert_pandas_to_tf_tensor(df[["X1", "X2"]]).shape
+        TensorShape([3, 2])
+
+        >>> from ray.data.extensions import TensorArray
+        >>>
+        >>> df = pd.DataFrame({"image": TensorArray(np.zeros((4, 3, 32, 32)))})
+        >>> convert_pandas_to_tf_tensor(df).shape
+        TensorShape([4, 3, 32, 32])
+    """
+
+    def tensorize(series):
+        try:
+            return tf.convert_to_tensor(series)
+        except ValueError:
+            return tf.stack([tensorize(element) for element in series])
+
+    try:
+        # We need to cast the tensors to a common type so that we can concatenate them.
+        # If the columns contain different types (for example, `float32`s and
+        # `int32`s), then `tf.concat` raises an error.
+        common_type: np.dtype = np.find_common_type(df.dtypes, [])
+        df = df.astype(common_type)
```

Comment:
Since this cast will often copy the DataFrame, which will then be (potentially) copied again once converting to a TensorFlow tensor, I think that it might be a better idea to do this cast when constructing the TensorFlow tensor.

```python
dtype: np.dtype = np.find_common_type(df.dtypes, [])
common_tf_dtype = tf.dtypes.as_dtype(dtype)

# ...

tensor = tensorize(series, dtype)

def tensorize(series, dtype):
    try:
        return tf.convert_to_tensor(series, dtype=dtype)
    except ValueError:
        return tf.stack([tensorize(element, dtype) for element in series])
```